app/middleware/headercert/middleware_test.go

```diff
@@ -0,0 +1,253 @@
+				},
+			},
+			&action{
+				status: http.StatusBadRequest,
+			},
+		),
+		gen(
+			"fail to verify the client cert",
+			[]string{},
+			[]string{},
+			&condition{
+				method: http.MethodGet,
+				headers: map[string]string{
+					"X-SSL-Client-Cert":        base64.StdEncoding.EncodeToString(failCert), // created by another rootCA.
+					"X-SSL-Client-Fingerprint": string(failFp),
+				},
+			},
+			&action{
+				status: http.StatusUnauthorized,
+			},
+		),
+		gen(
+			"fingerprint not matched",
+			[]string{},
+			[]string{},
+			&condition{
+				method: http.MethodGet,
+				headers: map[string]string{
+					"X-SSL-Client-Cert":        base64.StdEncoding.EncodeToString(cert),
+					"X-SSL-Client-Fingerprint": "invalid fingerprint",
+				},
+			},
+			&action{
+				status: http.StatusUnauthorized,
+			},
+		),
+		gen(
+			"expired client cert",
+			[]string{},
+			[]string{},
+			&condition{
+				method: http.MethodGet,
+				headers: map[string]string{
+					"X-SSL-Client-Cert":        base64.StdEncoding.EncodeToString(expiredCert), // expired client cert
+					"X-SSL-Client-Fingerprint": string(expiredFp),
+				},
+			},
+			&action{
+				status: http.StatusUnauthorized,
+			},
+		),
+	}
+
+	testutil.Register(table, testCases...)
+
+	rootCAs := []string{rootCAPath}
+	roots := x509.NewCertPool()
+
+	// Read the root cert specified in the local file
+	for _, c := range rootCAs {
+		rootCertPEM, err := os.ReadFile(c)
+		if err != nil {
+			t.Fatalf("fail to load rootCA: %v", err)
+		}
+		// Add the root cert to CertPool
+		if !roots.AppendCertsFromPEM(rootCertPEM) {
+			t.Fatalf("fail to load rootCA: %v", err)
+		}
+	}
+
+	opts := x509.VerifyOptions{
+		Roots: roots,
+	}
+
+	for _, tt := range table.Entries() {
+		tt := tt
+		t.Run(tt.Name(), func(t *testing.T) {
+
+			// Prepare the headercert middleware
+			headerCertMiddleware := &headerCert{
+				eh:   utilhttp.GlobalErrorHandler(utilhttp.DefaultErrorHandlerName),
+				opts: opts,
+			}
+
+			// Create a test request
+			req := httptest.NewRequest(tt.C().method, "http://test.com", nil)
+			for k, v := range tt.C().headers {
+				req.Header.Set(k, v)
+			}
+
+			// Create a test respose recoder
+			resp := httptest.NewRecorder()
+
+			// Call the middleware
+			headerCertMiddleware.Middleware(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusOK)
+			})).ServeHTTP(resp, req)
+
+			// verify the status code
+			testutil.Diff(t, tt.A().status, resp.Code)
+		})
+	}
+}
+
+func TestConvertCert(t *testing.T) {
+
+	t.Run("invalid pem", func(t *testing.T) {
+		invalidPEM := base64.StdEncoding.EncodeToString([]byte("-----BEGIN cert-----\nInvalid cert content"))
+		_, err := parseCert(invalidPEM)
+		if err == nil {
+			t.Fatal("expected error, got nil")
+		}
+	})
+
+	invalidCert, err := os.ReadFile(incompleteCertPath)
+	if err != nil {
+		t.Fatalf("fail to read client cert: %v", err)
+	}
+	t.Run("invalid x509 cert", func(t *testing.T) {
+
```

Comment:
L241-L244 should come here.

```suggestion
	invalidCert, err := os.ReadFile(incompleteCertPath)
	if err != nil {
		t.Fatalf("fail to read client cert: %v", err)
	}
```